File: quilt-ui/src/shared/utils/naturalDate.ts
// ──── NL Dates V1 — quilt-feature-natural-dates-v1 ────────────
//
// Convert natural-language date tokens ("today", "tomorrow",
// "yesterday") to ISO YYYY-MM-DD strings for use in property values.
//
// V1 scope: pure frontend transform. The user types `deadline:: today`
// in the inline editor; before persistence we rewrite it to
// `deadline:: 2026-06-05` so the backend always sees a real ISO date.
//
// V2 (not this file) would extend the DSL parser with `@today`,
// `@tomorrow` literals so the *raw* content preserves the intent and
// the resolver runs at query time. V1 keeps the resolver at the UI
// boundary — simpler, no schema migration, no DSL changes.
//
// Why a dedicated module (not inline in BlockRow.saveToApi)?
//   - It's pure: easy to unit-test deterministically with a `refDate`.
//   - It's the kind of helper that grows over time (next week, +3d…).
//   - Keeps the inline content layer (InlineContent.tsx) free of date
//     arithmetic, and the property rendering free of format quirks.
//
// All public functions are pure — no globals, no Date.now() at
// import-time, no timezone juggling beyond "local" components.
// That last point is deliberate: "today" means *the user's local
// today*, not UTC today, so someone in PST typing at 11pm shouldn't
// see tomorrow's date appear in their property.

/** Canonical property keys whose values are dates. Lowercased. */
const DATE_PROPERTY_KEYS = ['deadline', 'scheduled', 'date'] as const

/**
 * Resolve a single natural-language date token to a YYYY-MM-DD string.
 *
 * Accepts (case-insensitive, whitespace-trimmed):
 *   - `"today"`     → the reference date
 *   - `"tomorrow"`  → reference date + 1 day
 *   - `"yesterday"` → reference date − 1 day
 *
 * Returns `null` for any other input — including real ISO dates
 * (`"2026-01-15"`), empty strings, and compound tokens
 * (`"today!"`, `"today afternoon"`). The strict equality is what lets
 * us safely scan inline content without rewriting free-text mentions.
 *
 * @param input  The token to resolve. Typically a single word.
 * @param refDate  The "now" reference. Defaults to the current
 *                 `new Date()`. Tests pass a fixed date for
 *                 determinism — never read the wall clock in tests.
 */
export function resolveNaturalDate(
  input: string,
  refDate?: Date,
): string | null {
  if (typeof input !== 'string') return null
  const token = input.trim().toLowerCase()
  if (!token) return null

  if (token !== 'today' && token !== 'tomorrow' && token !== 'yesterday') {
    return null
  }

  const base = refDate ?? new Date()
  // `setDate` handles month/year overflow naturally
  // (setDate(0) → last day of previous month,
  //  setDate(32) on Dec 31 → Jan 1 of next year).
  const offsetDays = token === 'tomorrow' ? 1 : token === 'yesterday' ? -1 : 0
  const date = new Date(base)
  date.setDate(date.getDate() + offsetDays)

  return formatYmd(date)
}

/** Format a Date as `YYYY-MM-DD` using *local* components.
 *  Local (not UTC) is correct here because the user thinks in local
 *  time and "today" means local today. */
function formatYmd(d: Date): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

/**
 * Type guard: is this property key one that holds a date value?
 *
 * Recognises `deadline`, `scheduled`, and `date` (case-insensitive).
 * These are the keys the slash command menu already inserts
 * (see `makePropertyHandler` in slashRegistry.tsx).
 */
export function isDatePropertyKey(key: string): boolean {
  if (typeof key !== 'string') return false
  return (DATE_PROPERTY_KEYS as readonly string[]).includes(
    key.trim().toLowerCase(),
  )
}

/**
 * Scan a block's inline content and resolve any natural-date tokens
 * that appear as values of date properties (`deadline::`, `scheduled::`,
 * `date::`).
 *
 * Behaviour:
 *   - Only rewrites values that are *exactly* a natural-date token
 *     (post-trim). Free-text mentions like `"see you tomorrow!"` are
 *     left untouched.
 *   - Only rewrites values inside a known date property. `status::`
 *     stays a string.
 *   - Preserves surrounding whitespace, leading indentation, and
 *     other lines in multi-line content.
 *
 * This is the function BlockRow.saveToApi calls before `api.updateBlock`.
 *
 * @param content  The raw block content (the same string the user
 *                 typed in the contentEditable).
 * @param refDate  Optional "now" for determinism. Defaults to
 *                 `new Date()`.
 */
export function resolveNaturalDatesInContent(
  content: string,
  refDate?: Date,
): string {
  if (!content) return content

  return content
    .split('\n')
    .map(line => resolveNaturalDateInLine(line, refDate))
    .join('\n')
}

/** Per-line transform. Splits out for readability and easy testing. */
function resolveNaturalDateInLine(line: string, refDate?: Date): string {
  // Capture: 1=leading-ws, 2=key, 3=sep-ws, 4=value, 5=trailing-ws.
  // The value is `[^\s][^\n]*?` (starts with non-ws, non-greedy until
  // the trailing-whitespace+end anchor) so we don't eat the
  // separator's whitespace.
  const m = line.match(
    /^(\s*)(deadline|scheduled|date)::(\s*)([^\s][^\n]*?)(\s*)$/i,
  )
  if (!m) return line

  const [, leading, key, sepWs, value, trailingWs] = m
  const resolved = resolveNaturalDate(value, refDate)
  if (resolved === null) return line

  return `${leading}${key}::${sepWs}${resolved}${trailingWs}`
}

File: crates/quilt-domain/src/value_objects/natural_date.rs
//! Natural-language date resolution (V1).
//!
//! Mirrors the frontend `naturalDate.ts` module. Converts tokens like
//! "today", "tomorrow", "yesterday" to `YYYY-MM-DD` strings using the
//! *local* timezone of the server process.
//!
//! Scope: only rewrites exact single-token matches inside known date
//! property inline syntax (`deadline::`, `scheduled::`, `date::`).

use chrono::Local;

/// Resolve a single natural-language date token to `YYYY-MM-DD`.
///
/// Returns `None` for anything that isn't an exact match of
/// "today", "tomorrow", or "yesterday" (case-insensitive, trimmed).
pub fn resolve_natural_date(input: &str) -> Option<String> {
    let token = input.trim().to_lowercase();
    if token.is_empty() {
        return None;
    }

    let offset_days = match token.as_str() {
        "today" => 0,
        "tomorrow" => 1,
        "yesterday" => -1,
        _ => return None,
    };

    let base = Local::now().date_naive();
    let target = base + chrono::Duration::try_days(offset_days).unwrap();
    Some(target.format("%Y-%m-%d").to_string())
}

/// Canonical property keys whose values are dates.
const DATE_PROPERTY_KEYS: &[&str] = &["deadline", "scheduled", "date"];

/// Scan block content and resolve natural-date tokens in date-property
/// inline syntax (`deadline:: today` → `deadline:: 2026-06-12`).
///
/// Only rewrites values that are exactly a natural-date token.
/// Free-text mentions are left untouched.
pub fn resolve_natural_dates_in_content(content: &str) -> String {
    content
        .lines()
        .map(|line| resolve_natural_date_in_line(line))
        .collect::<Vec<_>>()
        .join("\n")
}

fn resolve_natural_date_in_line(line: &str) -> String {
    // Find the "::" separator (property inline syntax)
    let sep_pos = match line.find("::") {
        Some(p) => p,
        None => return line.to_string(),
    };

    // Extract the key (text before "::")
    let key_part = line[..sep_pos].trim_end();
    // Check if key is a known date property (case-insensitive)
    let key_lower = key_part.to_lowercase();
    if !DATE_PROPERTY_KEYS.contains(&key_lower.as_str()) {
        return line.to_string();
    }

    // Extract leading whitespace before the key
    let leading_ws_len = key_part.len() - key_part.trim_start().len();
    let leading_ws = &line[..leading_ws_len];

    // Extract the value after "::"
    let after_sep = &line[sep_pos + 2..];
    // Separate separator whitespace, value, and trailing whitespace
    let value_start = after_sep.find(|c: char| !c.is_whitespace());
    let (sep_ws, rest) = match value_start {
        Some(i) => (&after_sep[..i], &after_sep[i..]),
        None => return line.to_string(), // "::" with nothing after — no value to resolve
    };

    // Trim trailing whitespace from the value
    let trailing_ws_len = rest.len() - rest.trim_end().len();
    let value = &rest[..rest.len() - trailing_ws_len];
    let trailing_ws = &rest[rest.len() - trailing_ws_len..];

    match resolve_natural_date(value) {
        Some(resolved) => {
            format!("{}{}::{}{}{}", leading_ws, key_part, sep_ws, resolved, trailing_ws)
        }
        None => line.to_string(),
    }
}

/// Resolve a natural date value in a property write context.
/// If the key is a date property and the value is a natural token,
/// return the resolved ISO date. Otherwise return the value as-is.
pub fn maybe_resolve_date_property(key: &str, value: &str) -> String {
    let key_lower = key.trim().to_lowercase();
    if DATE_PROPERTY_KEYS.contains(&key_lower.as_str()) {
        match resolve_natural_date(value) {
            Some(resolved) => resolved,
            None => value.to_string(),
        }
    } else {
        value.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_today() {
        let result = resolve_natural_date("today").unwrap();
        let expected = Local::now().date_naive().format("%Y-%m-%d").to_string();
        assert_eq!(result, expected);
    }

    #[test]
    fn resolve_tomorrow() {
        let result = resolve_natural_date("tomorrow").unwrap();
        let expected =
            (Local::now().date_naive() + chrono::Duration::try_days(1).unwrap())
                .format("%Y-%m-%d")
                .to_string();
        assert_eq!(result, expected);
    }

    #[test]
    fn resolve_yesterday() {
        let result = resolve_natural_date("yesterday").unwrap();
        let expected =
            (Local::now().date_naive() + chrono::Duration::try_days(-1).unwrap())
                .format("%Y-%m-%d")
                .to_string();
        assert_eq!(result, expected);
    }

    #[test]
    fn case_insensitive() {
        assert!(resolve_natural_date("TODAY").is_some());
        assert!(resolve_natural_date("Tomorrow").is_some());
        assert!(resolve_natural_date("YESTERDAY").is_some());
    }

    #[test]
    fn whitespace_trimmed() {
        assert!(resolve_natural_date("  today  ").is_some());
    }

    #[test]
    fn rejects_non_tokens() {
        assert!(resolve_natural_date("foo").is_none());
        assert!(resolve_natural_date("next monday").is_none());
        assert!(resolve_natural_date("in 3 days").is_none());
        assert!(resolve_natural_date("today!").is_none());
        assert!(resolve_natural_date("not today").is_none());
    }

    #[test]
    fn rejects_empty() {
        assert!(resolve_natural_date("").is_none());
        assert!(resolve_natural_date("   ").is_none());
    }

    #[test]
    fn rejects_iso_dates() {
        assert!(resolve_natural_date("2026-01-15").is_none());
    }

    #[test]
    fn content_resolves_inline_deadline() {
        let result = resolve_natural_dates_in_content("deadline:: today");
        let expected = Local::now().date_naive().format("%Y-%m-%d").to_string();
        assert!(result.contains(&expected));
        assert!(result.starts_with("deadline::"));
    }

    #[test]
    fn content_preserves_non_date_properties() {
        let result = resolve_natural_dates_in_content("status:: todo priority:: A");
        assert_eq!(result, "status:: todo priority:: A");
    }

    #[test]
    fn content_preserves_real_dates() {
        let result = resolve_natural_dates_in_content("deadline:: 2026-01-15");
        assert_eq!(result, "deadline:: 2026-01-15");
    }

    #[test]
    fn content_preserves_free_text() {
        let result = resolve_natural_dates_in_content("I will finish this today");
        assert_eq!(result, "I will finish this today");
    }

    #[test]
    fn content_multiline_mixed() {
        let input = "some text\ndeadline:: today\nmore text";
        let result = resolve_natural_dates_in_content(input);
        assert!(result.starts_with("some text\n"));
        assert!(result.contains("deadline::"));
        assert!(result.ends_with("\nmore text"));
    }

    #[test]
    fn maybe_resolve_date_property_deadline_today() {
        let result = maybe_resolve_date_property("deadline", "today");
        let expected = Local::now().date_naive().format("%Y-%m-%d").to_string();
        assert_eq!(result, expected);
    }

    #[test]
    fn maybe_resolve_non_date_property_unchanged() {
        let result = maybe_resolve_date_property("status", "today");
        assert_eq!(result, "today");
    }

    #[test]
    fn maybe_resolve_real_date_unchanged() {
        let result = maybe_resolve_date_property("deadline", "2026-01-15");
        assert_eq!(result, "2026-01-15");
    }
}
